geocoder/hierarchy.cpp

```diff
@@ -1,13 +1,147 @@
+bool Hierarchy::Entry::DeserializeFromJSON(string const & jsonStr)
+{
+  try
+  {
+    my::Json root(jsonStr.c_str());
+    DeserializeFromJSONImpl(root.get());
+    return true;
+  }
+  catch (my::Json::Exception const & e)
+  {
+    LOG(LWARNING, ("Can't parse entry:", e.Msg(), jsonStr));
+  }
+  return false;
+}
+
+// todo(@m) Factor out to geojson.hpp? Add geojson to myjansson?
+void Hierarchy::Entry::DeserializeFromJSONImpl(json_t * root)
+{
+  if (!json_is_object(root))
+    MYTHROW(my::Json::Exception, ("Not a json object."));
+
+  json_t * const properties = my::GetJSONObligatoryField(root, "properties");
+
+  FromJSONObject(properties, "name", m_name);
+  m_nameTokens.clear();
+  search::NormalizeAndTokenizeString(m_name, m_nameTokens);
+
+  json_t * const address = my::GetJSONObligatoryField(properties, "address");
+
+  for (auto const & e : kKnownLevels)
+  {
+    string const & levelKey = e.first;
+    string levelValue;
+    FromJSONObjectOptionalField(address, levelKey, levelValue);
+    if (levelValue.empty())
+      continue;
+
+    EntryType const type = e.second;
+    CHECK(m_address[static_cast<size_t>(type)].empty(), ());
+    search::NormalizeAndTokenizeString(levelValue, m_address[static_cast<size_t>(type)]);
+  }
+
+  for (size_t i = 0; i < static_cast<size_t>(Hierarchy::EntryType::Count); ++i)
+  {
+    if (!m_address[i].empty())
+      m_type = static_cast<Hierarchy::EntryType>(i);
+  }
+
+  // The name of the entry must coincide with the most detailed
+  // field in its address.
+  if (m_type == Hierarchy::EntryType::Count)
+  {
+    LOG(LWARNING,
+        ("No address in an hierarchy entry. Name:", m_name, "Name tokens:", m_nameTokens));
+    return;
+  }
+  if (m_nameTokens != m_address[static_cast<size_t>(m_type)])
+  {
+    LOG(LWARNING, ("Hierarchy entry name is not the most detailed field in its address. Name:",
+                   m_name, "Name tokens:", m_nameTokens, "Address:", m_address));
+  }
+}
+
+// Hierarchy ---------------------------------------------------------------------------------------
 Hierarchy::Hierarchy(string const & pathToJsonHierarchy)
 {
-  UNUSED_VALUE(pathToJsonHierarchy);
+  fstream fs(pathToJsonHierarchy);
+  string line;
+
+  while (getline(fs, line))
+  {
+    if (line.empty())
+      continue;
+
+    auto i = line.find(' ');
+    CHECK(i != string::npos, ());
+    int64_t encodedId;
+    CHECK(strings::to_any(line.substr(0, i), encodedId), ());
+    line = line.substr(i + 1);
+
+    Entry entry;
+    // todo(@m) We should really write uints as uints.
+    entry.m_osmId = osm::Id(static_cast<uint64_t>(encodedId));
+
+    CHECK(entry.DeserializeFromJSON(line), (line));
+    m_entries[entry.m_nameTokens].emplace_back(entry);
+  }
+}
+
+shared_ptr<vector<Hierarchy::Entry>> Hierarchy::GetEntries(
+    vector<strings::UniString> const & tokens) const
+{
+  auto it = m_entries.find(tokens);
+  if (it == m_entries.end())
+    return {};
+
+  return make_shared<vector<Entry>>(it->second);
```

Comment:
Possible better way is return const pointer to vector.

Reply:
done